### ISSUE TITLE
ScriptNode MK3

### ISSUE THREAD

**Issue by zeffii** [nortikin/sverchok#839]
We learnt a lot from MK1, MK2, iojson and Monads, i'm tempted to think about the interface for ScriptNode MK3, and features. For all intents and purposes it should be a way to define a full-on first-class node, but without some of the boilerplate that gets in the way of experimentation/prototyping.

- optional override of `self.draw()`
- define non socket properties (bools, enums, whatever..)
- define operators / callbacks / button actions
- define softmin/softmax/defaults/etc for properties / sockets

Ideally i'd like it to provide a simple MK1 like interface, or MK2 (whatever the preference) , but I'll focus on MK1-simulacra perhaps @ly29 will provide MK2-simulacra or introduce redux concepts like type annotations? we can have various loaders which share common functions.

maybe a nice mix is possible.

**Comment by ly29** [comment]
I am certainly partial but I think the redux syntax is particularly neat. The MK2 syntax I think, in retrospect, was a mistake.

We know what is possible, is mostly a question about deciding a sane syntax. I think function annotations are perfect since they provide a simple way to introspect the function with out running it.
What I don't like is the return values where I am certain open for suggestions.

SvRx syntax
```python
def draw(self, context, layout):
    # do custom drawing

@node_func(label="Name", draw=draw)
def add(x: Number = 0.0, y: Number = 1.0) -> [("res", Number)]:
    return x + y
```

- import text files, as load mechanism, provides a clean and safe mechanism also for introspecting the files
- working with external files

Properties I think adding a gang of socket with custom specific properties like cycles is a sane option/ where you specify a type in the annotation. However since that might break a lot of things, there are many places where we would have to change things I think it is out of scope for the project.

So that leaves some options to define properties in function annotations. While the below could work it would soon become to extremely long.
```python
@node_func
def generate_something(start: FloatProperty(default=1.0, soft_min=0.0,soft_max=10.0),...
    start + ... # more
```
 So think some shorthand is needed 
Float Int etc could be automatically inserted into the namespace for script, for real nodes it should  be explicit I think.
Float etc would be basically be
```python
Float = dict
Int = dict
```

```python

@node_func
def do_stuff(start: Float(default=10.0, max=10.0, min=0.0)
     #
```

**Comment by ly29** [comment]
Importing text files as modules from SvRx. This allows for investigation of the live module.

```python
class SvRxFinder(importlib.abc.MetaPathFinder):

    def find_spec(self, fullname, path, target=None):
        if fullname.startswith("SverchokRedux.nodes.script."):
            name = fullname.split(".")[-1]
            name_py = "{}.py".format(name)
            if name in bpy.data.texts:
                return importlib.util.spec_from_loader(fullname, SvRxLoader(name))
            elif name_py in bpy.data.texts:
                return importlib.util.spec_from_loader(fullname, SvRxLoader(name_py))

        elif fullname == "SverchokRedux.nodes.script":
            # load Module, right now uses real but empty module, will perhaps change
            pass

        return None


class SvRxLoader(importlib.abc.SourceLoader):

    def __init__(self, text):
        self._text = text

    def get_data(self, path):
        # here we could insert a series of standard import and definitions
        return bpy.data.texts[self._text].as_string()

    def get_filename(self, fullname):
        return "<bpy.data.texts[{}]>".format(self._text)


sys.meta_path.append(SvRxFinder())
```

**Comment by ly29** [comment]
One feature I really would like is be able to work with an external text editor, so reloading  the node reloads the external file.

**Comment by zeffii** [comment]
> Properties I think adding a gang of socket with custom specific properties like cycles is a sane option/ where you specify a type in the annotation.

I thought most code that forced incoming socket types to be a specific kind was nixed?

> So think some shorthand is needed 

I never write the initial code for an sn script by using the verbose template, i always use sv_lang_template.sn (the rewriter).

```python

 When we pick a set of sensible defaults for the properties/sockets that we provide i wouldn't be surprised if we end up writing nodes more in SN MK3 than adding new dedicated nodes

**Comment by zeffii** [comment]
(not finished editing the above post, check back later)

**Comment by zeffii** [comment]
as for this
```python
@node_func
def generate_something(start: FloatProperty(default=1.0, soft_min=0.0,soft_max=10.0),...
    start + ... # more
```
we can allow for shorthand in the variables of the FloatProperty, later transliterating them
```python
@node_func
def generate_something(
   start: FloatProperty(d=0.0, smin=0.0, smax=360.0),
   radius: FloatProperty(d=2.0, smin=0.1, smax=10.0)
) -> [("res", Number)]:
    # function code here

```

**Comment by zeffii** [comment]
perhaps useful to consider the OSL extra annotations (not supported by Blenders OSL loader but interesting none the less)

**Comment by ly29** [comment]
OSL is good thing to look at. 
The default could be take from the python default value. 
Anyway we could allow different ways as long as we generate a python source that can then be loaded. 
sv_lang -> python
Python shorthand -> Python 
Etc.

**Comment by ly29** [comment]
The thing I am most unhappy about is the naming of the results/outputs socket...

**Comment by zeffii** [comment]
> The thing I am most unhappy about is the naming of the results/outputs socket...

help me understand your concern.

**Comment by ly29** [comment]
There isn't any way to name a return value in clean way like the inputs using function annotations:

Consider
```python
def func(a: StartValue=5) -> int:
    ....
```
How to we name int?
In redux I used ```("name", type)``` which is very different from how you write inputs.

Perhaps something like. ```type("name")``` looks a bit cleaner instead of the tuple though.

**Comment by zeffii** [comment]
maybe the return value could always be `-> Out` where out is defined in the local namespace

**Comment by zeffii** [comment]
i don't know if that' s possible.. but unless there's some technical limitation, and we can get over the counter intuitive concept of defining the output first, then i'm OK with it.  If you think about C++ you always define the output type first.. then the function.. so it's not completely alien.

**Comment by ly29** [comment]
Most people aren't used to function annotations in python anyway so there is no strict reason to adhere to it. However: following standards is good unless it's not.

  I would like for it be standard python but unless there is a *nice* syntax that we can define there is no reason to adhere to it. Right now it is an very open discussion.

If we have to use a preprocessor anyway we can start dreaming a bit...

However keeping in real :) the simplest definition I think is the one below.
```python
@node_func
def demo(a: Int=0, b: uInt =1) -> x: Int, y: Int:
     pass
```
Then we simply pre process the source to legal python...
```python
@node_func
def demo(a: Int=0, b: uInt =1) ->[("x", Int), ("y", Int)]:
     pass
```
Which is simple enough to do...

**Comment by ly29** [comment]
Will try to implement load etc as I think about leaving the syntax part open for testing.

**Comment by zeffii** [comment]
i was also thinking about being able to limit values carried by incoming sockets, just having a max/min on the slider when unconnected is a little weak. The downside to forcing values to adhere to a range is the extra processing. in my pipe dream an incoming socket would look like 

```
    _________________________________________
   |___________ node name ___________________|
   |                                         0
   |                                         |
   0 sockname        [clamp][remap to limit] |
   0 sockname2       [clamp][remap to limit] |
   |_________________________________________|
```

**Comment by zeffii** [comment]
thinking out loud, not intended to be taken serious

**Comment by ly29** [comment]
I consider slider values more like softmax/softmin and it is the responsiblity of the user to get correct input, crashing and throwing an error is ok I think in most cases.

**Comment by ly29** [comment]
But such input verification could be provided.

**Comment by ly29** [comment]
First commit, mostly some infrastructure ported over from redux

**Comment by zeffii** [comment]
question..as am reading:
https://github.com/nortikin/sverchok/blob/754fda8c255de5de57eff19b10679870ae2ba8d4/utils/loadscript.py#L24-L25

do those really need to be global ? that function will always look in the outer namespace for variables, and dicts seem to act as pass-by-reference.

**Comment by ly29** [comment]
Looking back I think you are right, I need to review my knowledge of globals, the precise semantics.

**Comment by zeffii** [comment]
i think we may have a few of them :) and only needed in the case where globals refer to strings (or lists.too..i'm not 100% about that),

**Comment by zeffii** [comment]
so this is a valid node func?

```python
@node_func
def demo(a: Int=0, b: Int=1) ->[("x", Int), ("y", Int)]:
     pass
```

**Comment by zeffii** [comment]
```python
@prop_args(
    a=dict(smin=1, min=0, smax=20, default=4),
    a=dict(smin=1, min=-30, smax=20, default=14)
)
@node_func(...)
def demo(a: Int=0, b: Int=1) -> [("x", Int), ("y", Int)]:
     pass
```

**Comment by ly29** [comment]
@zeffii 
But you have defaults in two places now there

**Comment by zeffii** [comment]
corrected!

**Comment by ly29** [comment]
How about a special parameter called ```Node``` that will be a reference to the current node, for all sorts of trickery that we don't envision as being needed for the standard case?

To be clear I don't consider properties to fall into the this special trickery.
 
```
@node_func
def magical_function(a: Int = 3, b: uInt = 10, node: Node = None) -> [("x", Int]
    tree = node.id_data
   do_magic(tree)
```

**Comment by zeffii** [comment]
lets have a dedicated mk3 callback handler?
```python


```


```python
def draw(self, context):
    if self.my_bool:
        ...
    else:
        ...
        k = self.layout.operator('node.svrx_callback', text='do it!')
        k.func_name = 'do_this_thing'
        # k.func_prop1 = self.a    can also be inferred from the `self` value being passed to the operator

@prop_args(
    a=dict(smin=1, min=0, smax=20),
    b=dict(smin=1, min=-30, smax=20),
    mybool=dict(default=True)
)
@node_func(draw=draw)    # callback=callback (a generic operator which accept
def demo(a: Int=0, b: Int=1) -> [("x", Int), ("y", Int)]:
     pass

```

**Comment by zeffii** [comment]
are you intentionally calling that property a `uInt` (unsigned int?)

**Comment by ly29** [comment]
yeah, just thinking a bit out loud a bit. I am consider a gang of sockets for various types.

**Comment by zeffii** [comment]
makes total sense, even if the only nodes using the extra specific sockets is (for the time being) SNmk3

**Comment by zeffii** [comment]
not sure btw about getting the callback to find the right namespace for the custom function..

```python
   
master_sn_func_dict = {}

   # inside callback

   def execute(self, context):
      try:
            master_sn_func_dict.get(self.func_name)(node)
      except:
            ...

**Comment by ly29** [comment]
I think the sockets hierarchy probably is a new issue.

**Comment by zeffii** [comment]
edited the above comment

**Comment by ly29** [comment]
Re the namespace.

```
@property
def module(self):
    mod_name = self.func_name.__module__.split(".")[-1]
    return getattr(sverchok.nodes.scripts, mod_name)
```

```
def execute(self, context):
    node = context.node
    node.module.get(self.func_name)(node)
```

**Comment by zeffii** [comment]
still might want to pass `context` by default.. it's just a reference anyway
```python
node.module.get(self.func_name)(node, context)
```

in case the operator on the node is used to effect the current node tree somehow.

**Comment by ly29** [comment]
Yeah, context should be passed on.

**Comment by zeffii** [comment]
it seems you have most of it figured out, i'll find something else to do :)

**Comment by ly29** [comment]
I expect to have a basic version up and running tomorrow.
Then we can sort out details, decide syntax,  make examples etc.

**Comment by ly29** [comment]
I don't if I will have time today actually, will try. Almost everything is in place but have to tie it together and test that actually works like I think.

**Comment by ly29** [comment]
Got some time today, got it out of sketches and made it more concrete. still doesn't work yet but now the direction is clear.

**Comment by ly29** [comment]
![skarmavbild 2016-08-31 kl 09 56 55](https://cloud.githubusercontent.com/assets/6241382/18163710/a0d1f120-703c-11e6-9296-51d23a859422.png)
Success! 
Fragile, doesn't reload and many small issues but finally something to show.

**Comment by zeffii** [comment]
excellent, - time for a list of checkboxes..to keep track of the progress! and todo

**Comment by ly29** [comment]
Yeah, will try to get a check list up today, some issues of the top of my mind

- [ ] Allow ```Int``` instead of ```Int()```
- [ ] Properties
- [ ] Custom socket properties
- [ ] Test the custom class
- [ ] Decide syntax
- [ ] Workflow and UI
- [ ] Reload
- [ ] Load on startup

**Comment by ly29** [comment]
You want to plug in SN1 node scripts still? Would be good the keep in mind, haven't thought much about it now.

**Comment by zeffii** [comment]
yeah, I can write a preparser which would take normal SN1 scripts and augmented-SN1 for extra stuff.

**Comment by ly29** [comment]
@zeffii 
Let us try to fit it in quite early, generalizing the design is easier early when it is still flexible before starting trying to deal with corner cases...
Basiclly a different loader and some wrappers.

**Comment by zeffii** [comment]
take SN1's first script example..
https://github.com/nortikin/sverchok/blob/master/node_scripts/templates/petal_sine.py
or the much shorter
https://github.com/nortikin/sverchok/blob/master/node_scripts/templates/zeffii/vector_fx.py

petal_sine.py, often the sockets are almost redundant and could be inferred from the function signature..and now also the outputs.., and only in the case where the scripter really wants a different name to appear on the slider than the name of the variable to use inside the code, then we should offer a small mapping dict.

```python
from math import sin, cos, radians, pi
from mathutils import Vector, Euler
# creates flower

def sv_main(n_petals=8, vp_petal=20, profile_radius=1.3, amp=1.0):

    # in_sockets = [
    #     ['s', 'Nm Petals',  n_petals],
    #     ['s', 'herts per Petal',  vp_petal],
    #     ['s', 'Profile Radius', profile_radius],
    #     ['s', 'Amp',  amp],
    # ]

    stuff

    # out_sockets = [
    #     ['v', 'Verts', [Verts]],
    #     ['s', 'Edges', [Edges]],
    # ]

    return
```

**Comment by zeffii** [comment]
i've never liked defining the args on sv_main directly..
```python

from math import sin, cos, radians, pi
from mathutils import Vector, Euler
# creates flower

input_dict = ordered_dict(
    num_petals=(Int, min=0, max=20),
    herts_per_petal=(Int, min=0, max=20, show_name='herts'),
    ...
    verts_in=(VectorSocket, levels=2)
)

output_dict = ordered_dict(
    verts=(VectorSocket, varname=my_verts),   # v
    edges=(CollectionSocket),  # s
    ...
)


@nodefunc(....)
def sv_main(**args): -> output_dict

    return verts, edges
```

if something like this can be made to work, then it has a great change of entirely replacing my love for SN1.

**Comment by ly29** [comment]
I intend to support the different socket/name slider name like the following. Since a good name inside the function and the user interface name doesn't necessary match. The name parameter to the type is optional. Also intend for the same interface for custom types, but done it yet.

```python

@node_script
def petal(n_petals: uInt("Nm petals") = 8, 
               vp_petal: uInt("herts per Petal)=20, 
               profile_radius: Float('Profile Radius')=1.3, 
               amp: Float("Amp")=1.0) -> (Vertex("Verts"), Edge("Edges")):

```

**Comment by ly29** [comment]
Something like that could absolutely be supported, even though my preferences goes the other way.

**Comment by ly29** [comment]
Something like below the syntax might become.
```python
input_dict = ordered_dict(
    num_petals=Int(min=0, max=20),
    herts_per_petal=Int(min=0, max=20, show_name='herts'),
    ...
    verts_in=(VectorSocket, levels=2)
)

def f(*param: input_dict) -> output_dict:
    pass
```

**Comment by zeffii** [comment]
```
@node_script
def petal(
    n_petals: uInt("Nm petals") = 8, 
    vp_petal: uInt("herts per Petal)=20, 
    profile_radius: Float('Profile Radius')=1.3, 
    amp: Float("Amp")=1.0
) -> (
    Vertex("Verts"),
    Edge("Edges")
):

    # function here

    return Verts, Edges
```

i'd probably end up formatting it like that if that's not too complicated for the type system..

**Comment by ly29** [comment]
Yeah that should work perfectly I think.

**Comment by zeffii** [comment]
or maybe a class that returns an ordered dict somehow

i = SN_dict()
i.n_petals = uInt("Nm petals"), 8
i.vp_petal = uInt("herts per Petal), 20
i.profile_radius = Float('Profile Radius'), 1.3
i.amp: Float("Amp"), 1.0

**Comment by zeffii** [comment]
i ges with `__attrgetter__`  it could count how many items it has, and tag it...

**Comment by ly29** [comment]
The type system at moment is very extremely simple.
```python
class SvBaseType:
    def __init__(self, name=None):
        self.name = name

class Vertex(SvBaseType):
    bl_idname = "VerticesSocket"

class Int(SvBaseType):
    bl_idname = "SvIntSocket"

class uInt(SvBaseType):
    bl_idname = "SvUnsignedIntSocket"
```

**Comment by zeffii** [comment]
```python
from svsn3 import hide

i = SN_dict()
i.n_petals = uInt("Nm petals", 8)
i.vp_petal = uInt("herts per Petal, 20)
i.profile_radius = Float('Profile Radius', 1.3)
i.amp = Float("Amp", 1.0)
i.some_bool = Bool("Switch me", True), hide   # for use with `def draw()`
```

**Comment by ly29** [comment]
The properties and custom sockets have to support more and also, perhaps, be of a different type. 

```python
class Float: #name is possible to discuss
    # as property with value in socket if options aren't set
    bl_idname = "StringsSocket"
    def __init__(self, name=None, **options):
        self.name = name
        self.prop_dict = options
```
They could be unified but I am not sure that it is a good idea. It is good to be clear of what goes into the interface and what ends up as socket.

```python
class FloatP: #name is possible to discuss
    def __init__(self, name=None, **options):
        self.name = name
        self.prop_dict = options
```

I SvRx I had as rule that a certain order was required but now I am thinking that any order is okay but the type deciedes which is what. Property like they get automatically exposed in the interface but if you want to look nice you have to take care of with custom draw.

**Comment by zeffii** [comment]
nope, i'm OK Float vs FloatP if the the P means an internal prop..

unfortunately complexity must exist either in the backend (so it seems like magic to users ..and us after a while) or on the front.

**Comment by ly29** [comment]
I don't think I will make a SN2 compability mode, but I will port over any scripts upon request. My general feeling is that SN1 is much more used...
SN2 goes from beta to old node...

**Comment by zeffii** [comment]
```python
from collections import OrderedDict
hide = 'Hidden'

class uInt:
    def __init__(self, *args, **kw):
        ...

class Float:
    def __init__(self, *args, **kw):
        ...

class Bool:
    def __init__(self, *args, **kw):
        ...


class SN_dict:

    def __init__(self):
        self.__dict__ = OrderedDict()

    def __setattr__(self, key, val):
        self.__dict__[key] = val

    def __getattr__(self, key):
        return self.__dict__[key]
    
    def keys(self):
        return self.__dict__.keys()

    def value(self):
        return self.__dict__.values()

    def items(self):
        return zip(self.__dict__.keys(), self.__dict__.values())


i = SN_dict()
i.n_petals = uInt("Nm petals", 8)
i.vp_petal = uInt("herts per Petal", 20)
i.profile_radius = Float('Profile Radius', 1.3)
i.amp = Float("Amp", 1.0)
i.some_bool = Bool("Switch me", True), hide   # for use with `def draw()`


print(i.__dict__.items())
```

but have seen smarter people (I assume they're smarter) take much more verbose approaches:
http://stackoverflow.com/a/14048352/1243487

**Comment by ly29** [comment]
I am not sure I understand the advantage of this approach. You still have to unpack the dictionary, but I suppose we could provide the variables in closure in  ```node_script``` so ```n_petals``` is available in the namespace.

```python
I.some_bool =
@node_script
def g(p=inputs):
     if np_petals < 10:
         # do stuff
```

**Comment by zeffii** [comment]
there may be no advantage :)

**Comment by ly29** [comment]
I guess the below isn't too bad a syntax
```python
def g(*p:order_dict_settings) -> outputs:
     petal_nr, abc, ... = p
```

**Comment by zeffii** [comment]
```python
config = SN_dict()

with config.inputs as i:
    i.n_petals = uInt("Nm petals", 8)
    i.vp_petal = uInt("herts per Petal", 20)
    i.profile_radius = Float('Profile Radius', 1.3)
    i.amp = Float("Amp", 1.0)

config.props.some_bool = Bool("Switch me", True)   # for use with `def draw()`

with config.outputs as o:
    o.outputs.verts = ...
    o.outputs.edges = ...
```

**Comment by zeffii** [comment]
def g(*p:config.inputs) -> config.outputs:
     petal_nr, abc, ... = p

**Comment by zeffii** [comment]
i'd rather we slow down at this stage and throw away a lot of bad ideas than picking too soon.

**Comment by ly29** [comment]
Yeah, let us think. I won't have time to push forward much tomorrow anyway.

**Comment by ly29** [comment]
![skarmavbild 2016-09-02 kl 08 21 51](https://cloud.githubusercontent.com/assets/6241382/18194498/5f93ae06-70e6-11e6-9a43-5abea6cdb19c.png)
some progress. but now tuple unpacking broke. for some reason... 

```takes from 0 to 3 positional arguments but 4 were given```

**Comment by zeffii** [comment]
is it auto injecting self ?

**Comment by ly29** [comment]
That sounds like might happen, hmm. Good catch. 
Oh...  Makes total sense, I have store the function some other way and not in the class dictionary I guess.

**Comment by ly29** [comment]
Or apply staticmethod when it is put into the class dictionary.

**Comment by ly29** [comment]
I will try to push mostly with functionality to make things easy and clear. Unless the end result is easier and clearer than ```SN1/sv_main``` I think it is worthless.

The ```sv_lang``` templates could easily be used to auto generate the function signature.

**Comment by zeffii** [comment]
would it be undesirable to have the node's .self available inside the node_func ?

**Comment by ly29** [comment]
No, it is in fact supported if you use the ```Node``` type (but untested), but I think a ```SN1``` simplicity by default is good thing.

```python
@node_script
def g(self: Node, value_a : Float=1.0) -> (Float,)
    pass
```

**Comment by zeffii** [comment]
neat! it will be fun to write some templates for it. and maybe slowly convert regular nodes to this kind of style, We need some way to determine if it's worth processing certain parts of the code dependant on outputs being linked

would something like

```python
@node_script
def g(self: Node, value_a : Float=1.0) -> (Verts, Edges)

    if Node.outputs("Edges").is_linked():
        # then generate edges..

    return Verts, Edges
```

**Comment by ly29** [comment]
Yes, with that I agree, not sure about how to go about it in a clean way.

One option is the ```Node``` special argument and parse like above, another such argument could be a ```dict``` with ```{socket_name: Bool}``` or something along those lines. ```[(name, bool)]``` perhaps.

**Comment by zeffii** [comment]
or expose a local var..

@node_script
def g(self: Node, value_a : Float=1.0) -> (Verts, Edges)

    if Node.linked("Edges"):
        # then generate edges..

    return Verts, Edges

**Comment by zeffii** [comment]
but what if an input and an output share the same name, then `has_links()` would perhaps take a second parameter `has_links(name, iface='inputs/outputs/both')`

**Comment by ly29** [comment]
It is point worth exploring, but the basic syntax is growing on me. But I am somewhat partial.

In general a node shouldn't care if a input is linked or not, realistically one could use a ```sentinel``` value as default for cases when the input is truly optional, if it is needed just ask for it and show than you can be processed via the not ready coloring.
 So three inputs cases:
1. Socket has default value that can be used, either as supplied default or as a exposed property
2. Socket does not have a default and it is needed, think topology, in which case it gets that orange yellow color
3. Node tries to be somewhat clever and does things depending on if socket is linked or not. 
4. Or even worse, depending on what kind of socket it has linked...

1,2 and 3 should be supported in a good way. I think for 3 having a sentinel value as default and checking for it is a sane approach.

For 4 you are on your own. Nodes that are too clever are hard to use and lead to confusion, anger and bug reports.

This points are on my mind but for later stage than the first step, which is achieving SN1 like functionality. But they are worth discussing, also socket names aren't unique...

**Comment by ly29** [comment]
```python

@node_script
def vertex_color(obj: Object = None, index: Index = None, colors: Vertex = None)-> Vertex("Colors"):
    if obj is None or colors is None: 
         raise SomeError
    if index is None:
          #generate index
    # now we can process...
```

**Comment by zeffii** [comment]
4 is not something i'm worried about, certainly not for these nodes which are intended to be simple to write and augment.

there is something to consider, a scriptnode version of a Math node, at least conceptually one might want to be able to have StringProperty and EnumProperty and depending on the chosed function have a diverse number of inputs outputs...  or maybe such dynamic nodes would be too convoluted for this system.

**Comment by zeffii** [comment]
may be beyond scriptnode territory..

**Comment by ly29** [comment]
Well, not convoluted per se but somewhat more complicated.

I can imagine some ways of doing it, at least conceptually, without a clear obvious way. Right now it is a bit beyond the scope of what I try to achieve.

Looking at math node and such with the amount of support code compared to the simple actual operation...

```python
@multi_node_func
def mode_picker(enum: EnumP): -> NodeFunc
       return node_func(enum)

def node_func(enum):
      op_typ = operations[enum]
      def fxy(x: Float=0, y: Float) -> Float:
            do_everything(op_typ)
      def fx_x(...
      def fx_xy(...
      def constant()...
      if op_typ in constants:
           return constant
      etc...
```




```

**Comment by zeffii** [comment]
well, settle on the more generic MK3 syntax first, then after that it's worth considering extensions -- not exactly high up on the todo.

and the MK3 needs to be serializable :)

**Comment by ly29** [comment]
This is preview
```python

# SN3 Syntax description
# @node_script is needed for to declare at a function will be the main function of the node

@node_script
def f(...
# node script takes a series of optional parameters like label which definies the node label.

# to describe the inputs and outputs of the node the python function annotation is used
@node_script
def f(a: Int=1, b: Float=1.0) -> (Float, Int("Count"):
     # function code
# Input parameters
# for each parameter three things are needed
name: Type = default
# name is the normal variable name as used in the function
# Type one of the following:
# -  is the socket type decleared using a set of possible types providing convenient properties and named, the main option being the first parameter to the type is the name of the socket, which can be set. otherwise a default value is used 
# - a blender property decleared using a shorthand like FloatP, IntP, 
# - Node, will give a reference to the node. this is an advanced option
# default will be used

# Note, for more advanced properties/types it can be declared before the function

my_float = FloatP(lots options)

@node_script
def f(a: my_float=1.0) -> ...

# Results
```
To be continued

**Comment by ly29** [comment]
I think I want to continue this more in two new issues. As many things have been discussed back and forth it is hard to follow.

1. SN3 Syntax, and discussion, the main remaining issue is if one could write the inputs outside of the function header. Also some examples.
2. SN3 Implementation and issues.

**Comment by ly29** [comment]
Also the sockets part could be pushed to master earlier, since more socket refactoring will happen soonish with object socket etc.

**Comment by zeffii** [comment]
ok, i'm going to start doing some geom implementations.

**Comment by ly29** [comment]
![skarmavbild 2016-09-04 kl 11 25 07](https://cloud.githubusercontent.com/assets/6241382/18230130/4a9fe0ac-7292-11e6-95e9-f47dd6aca5ed.png)

Test with custom drawing and string property
```python
import bpy
import numpy as np


"""
Sample an image at u,v coordinates, repeatin the texture
Images can either be selected by name or int.
Outputs image data at uv as rgb
"""

def draw(self, context, layout):
    layout.prop_search(self, "image_name", bpy.data, "images")

@node_script(draw_buttons=draw)
def sample_image(image_name: StringP() = "",
                 points: Vertex("Uv points") = None) -> (
                 Vertex("Colors"),
                 Float("Alpha")):

    bpy_image = bpy.data.images[image_name]

    dim_x, dim_y = bpy_image.size
    tmp = np.array(bpy_image.pixels)
    image = tmp.reshape(dim_y, dim_x, 4)

    out = []
    alpha = []
    points = np.array(points[0])
    uv = points * (dim_x, dim_y, 0)
    #
    for u, v, _ in uv:
        color = image[v % dim_y, u % dim_x]
        out.append(color[:3].tolist())
        alpha.append(color[-1])


    return out, alpha

```

**Comment by ly29** [comment]
So that is also the first script that uses SN3 syntax and actually does something, I realized.

**Comment by ly29** [comment]
@zeffii 
How do you feel about the syntax?

**Comment by zeffii** [comment]
I am pleased with this! It think people will get their head around it because it is sufficiently declarative, it also remains true to the abstraction of SN1. We make the function the most important element, and leave the boilerplate to what can be inferred by the machine (or defaults).

In the end we can't hold people's hands all the way..

**Comment by ly29** [comment]
see other isses for more discussion
